[pytorch][codegen] gen_python_functions.py loading native_functions.yaml / deprecated.yaml directly

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #47818 [pytorch][codegen] migrate gen_variable_factories.py to the new data model
* #47747 [pytorch][codegen] add fully migrated scripts to mypy strict config
* **#47746 [pytorch][codegen] gen_python_functions.py loading native_functions.yaml / deprecated.yaml directly**
* #47745 [pytorch][codegen] migrate gen_annotated_fn_args.py to new codegen model

- Removed the integration hack in gen_python_functions.py. It now directly
  loads native_functions.yaml. All dependencies on Declarations.yaml
  have been removed / moved to elsewhere.
- Rewrote the deprecated.yaml parsing logic to work with new data model directly.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24885067](https://our.internmc.facebook.com/intern/diff/D24885067)

ljk53 committed:
[pytorch][codegen] gen_python_functions.py loading native_functions.yaml / deprecated.yaml directly

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][codegen] gen_python_functions.py loading native_functions.yaml / deprecated.yaml directly"


- Removed the integration hack in gen_python_functions.py. It now directly
  loads native_functions.yaml. All dependencies on Declarations.yaml
  have been removed / moved to elsewhere.
- Rewrote the deprecated.yaml parsing logic to work with new data model directly.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24885067](https://our.internmc.facebook.com/intern/diff/D24885067)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][codegen] gen_python_functions.py loading native_functions.yaml / deprecated.yaml directly"


- Removed the integration hack in gen_python_functions.py. It now directly
  loads native_functions.yaml. All dependencies on Declarations.yaml
  have been removed / moved to elsewhere.
- Rewrote the deprecated.yaml parsing logic to work with new data model directly.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24885067](https://our.internmc.facebook.com/intern/diff/D24885067)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][codegen] gen_python_functions.py loading native_functions.yaml / deprecated.yaml directly"


- Removed the integration hack in gen_python_functions.py. It now directly
  loads native_functions.yaml. All dependencies on Declarations.yaml
  have been removed / moved to elsewhere.
- Rewrote the deprecated.yaml parsing logic to work with new data model directly.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24885067](https://our.internmc.facebook.com/intern/diff/D24885067)

[ghstack-poisoned]

ezyang commented on tools/autograd/gen_python_functions.py:
```diff
@@ -115,230 +107,217 @@ def should_generate_py_binding(f: NativeFunction) -> bool:
 
-    return {
+    fm.write_with_template(filename, filename, lambda: {
+        'generated_comment': f'@generated from {fm.template_dir}/{filename}',
```
Don't do this, now editors think this file is generated. '@' + 'generated' or something similar

ljk53 replied:
done

ezyang commented:
I'll review this after @generated fix, as the improper token is thwarting Phabricator's code movement detection

ljk53 committed:
Update on "[pytorch][codegen] gen_python_functions.py loading native_functions.yaml / deprecated.yaml directly"


- Removed the integration hack in gen_python_functions.py. It now directly
  loads native_functions.yaml. All dependencies on Declarations.yaml
  have been removed / moved to elsewhere.
- Rewrote the deprecated.yaml parsing logic to work with new data model directly.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24885067](https://our.internmc.facebook.com/intern/diff/D24885067)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch][codegen] gen_python_functions.py loading native_functions.yaml / deprecated.yaml directly"


- Removed the integration hack in gen_python_functions.py. It now directly
  loads native_functions.yaml. All dependencies on Declarations.yaml
  have been removed / moved to elsewhere.
- Rewrote the deprecated.yaml parsing logic to work with new data model directly.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24885067](https://our.internmc.facebook.com/intern/diff/D24885067)

[ghstack-poisoned]

ljk53 commented on tools/autograd/gen_python_functions.py:
```diff
@@ -115,230 +107,218 @@ def should_generate_py_binding(f: NativeFunction) -> bool:
+        opname = str(f.func.name.name.base)
+        if f.func.is_out_fn():
+            opname += '_out'
+        args = CppSignatureGroup.from_schema(f.func, method=False).signature.arguments()
```
@bhosmer updated this according to our offline discussion - it turns out that a.argument could be TensorOptionsArguments (and I cannot simply access `a.argument.type` because of it).

bhosmer reviewed (approved):
Excellent!

ljk53 committed:
Update on "[pytorch][codegen] gen_python_functions.py loading native_functions.yaml / deprecated.yaml directly"


- Removed the integration hack in gen_python_functions.py. It now directly
  loads native_functions.yaml. All dependencies on Declarations.yaml
  have been removed / moved to elsewhere.
- Rewrote the deprecated.yaml parsing logic to work with new data model directly.

Confirmed byte-for-byte compatible with the old codegen:
```
Run it before and after this PR:
  .jenkins/pytorch/codegen-test.sh <baseline_output_dir>
  .jenkins/pytorch/codegen-test.sh <test_output_dir>

Then run diff to compare the generated files:
  diff -Naur <baseline_output_dir> <test_output_dir>
```

Differential Revision: [D24885067](https://our.internmc.facebook.com/intern/diff/D24885067)

[ghstack-poisoned]
